crates/slightfile/src/capability_store.rs

```diff
@@ -0,0 +1,177 @@
+    fn as_mut(&mut self) -> &mut HashMap<CapabilityName, HashMap<ResourceName, T>> {
+        &mut self.inner
+    }
+}
+
+impl<T> From<HashMap<CapabilityName, HashMap<ResourceName, T>>> for CapabilityStore<T> {
+    fn from(inner: HashMap<CapabilityName, HashMap<ResourceName, T>>) -> Self {
+        Self { inner }
+    }
+}
+
+impl<T> From<CapabilityStore<T>> for HashMap<CapabilityName, HashMap<ResourceName, T>> {
+    fn from(val: CapabilityStore<T>) -> Self {
+        val.inner
+    }
+}
+
+impl<T> std::ops::Deref for CapabilityStore<T> {
+    type Target = HashMap<CapabilityName, HashMap<ResourceName, T>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl<T> std::ops::DerefMut for CapabilityStore<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<T> std::iter::FromIterator<(CapabilityName, HashMap<ResourceName, T>)> for CapabilityStore<T> {
+    fn from_iter<I: IntoIterator<Item = (CapabilityName, HashMap<ResourceName, T>)>>(
+        iter: I,
+    ) -> Self {
+        Self {
+            inner: iter.into_iter().collect(),
+        }
+    }
+}
+
+impl<T> std::iter::IntoIterator for CapabilityStore<T> {
+    type Item = (CapabilityName, HashMap<ResourceName, T>);
+    type IntoIter = std::collections::hash_map::IntoIter<CapabilityName, HashMap<ResourceName, T>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.inner.into_iter()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::Resource;
+    use anyhow::Result;
+
+    use super::*;
+
+    #[test]
+    fn test_capability_store() -> Result<()> {
+        #[derive(Debug, Clone, Eq, PartialEq)]
+        struct State {
+            implementor: Resource,
+            name: String,
+        }
+
+        let state_1 = State {
+            implementor: Resource::Keyvalue(crate::resource::KeyvalueResource::Redis),
+            name: "my-container".into(),
+        };
+
+        let state_2 = State {
+            implementor: Resource::Messaging(crate::resource::MessagingResource::Mosquitto),
+            name: "my-pubsub".into(),
+        };
+
+        let state_3 = State {
+            implementor: Resource::Keyvalue(crate::resource::KeyvalueResource::Filesystem),
+            name: "my-other-container".into(),
+        };
+
+        let mut store = CapabilityStore::new();
+
+        store.insert(
+            ResourceName::Specific("my-container".to_owned()),
+            "keyvalue",
+            state_1.clone(),
+        );
+        store.insert(ResourceName::Any, "messaging", state_2.clone());
+        store.insert(
+            ResourceName::Specific("my-other-container".to_owned()),
+            "keyvalue",
+            state_3.clone(),
+        );
+
+        assert_eq!(store.get("my-container", "keyvalue"), Some(&state_1));
+        assert_eq!(store.get("my-pubsub", "messaging"), Some(&state_2));
+        assert_eq!(store.get("my-other-container", "keyvalue"), Some(&state_3));
```

Comment:
Your examples are not quite aligned. In line 29, you had `store.get(.)` take 1 parameter – a specific resource. Here, you're passing the name `first` (i.e., not `resource`), and a non-specific capability `resource`.

Also, in line 51, you used `name` in `get` as the value to finding a `ResourceName::Specific`. How's that if `my-container` is `name`?

In other examples (i.e., line 29), you used values like `keyvalue.redis` as the value to `ResourceName::Specific`.

Reply:
Thanks for pointing it out! Line 29 is out-of-date and I forgot to update it.

Also, in line 51, you used name in get as the value to finding a ResourceName::Specific. How's that if my-container is name?

I am not quiet sure if I understand the question. my-container is passed as name parameter to get function.

Reply:
> I am not quiet sure if I understand the question. my-container is passed as name parameter to get function.

My confusion is w/ the fact that you're using name in line 51 w/ conjunction to `ResourceName::Specific(.)` (e.g., `ResourceName::Specific("my-container")`; while, in other examples (i.e., line 26), you used `ResourceName::Specific(.)` like `ResourceName::Specific("keyvalue.redis")`, which are two different things, but I'm guessing that's just 'cause the doc comments are old.

Reply:
I fixed the docs. Does it answer your questions now?